src/proposal.jl

```diff
@@ -0,0 +1,18 @@
+abstract type AbstractProposal{T<:AbstractTrajectory} end
```

Comment:
Perhaps consider merging the `AbstractProposal` and `AbstractTrajctory` types by: 

```julia
abstract type AbstractProposal end
abstract type AbstractHamiltonianTrajectory <: AbstractProposal end
abstract type StaticTrajectory <: AbstractHamiltonianTrajectory end
abstract type NoUTurnTrajectory <: AbstractHamiltonianTrajectory end
abstract type UniformProposal <: NoUTurnTrajectory
abstract type MultinomialProposal <: NoUTurnTrajectory
abstract type TakeLastProposal <: StaticTrajectory # this type might be unnecessary since we always take the last point for static trajectory

# then define `propose` for each `t`
propose(t::AbstractHamiltonianTrajectory, ...) 
```

This design of types might be easier to reason about since intuitively Hamiltonian simulation is a way to make MCMC proposals. It also allow us to extend `AbstractProposal` with non-Hamiltonian proposals.

Reply:
> this type might be unnecessary since we always take the last point for static trajectory
Is there benefit of using a multinomial sampling instead?

Reply:
@yebai Does using a multinomial sampling from static trajectory make sense?

Reply:
I'd like to merge them if arbitrary combination of sampling and ways of building trajectory doesn't make (practical) sense.

Reply:
> Is there benefit of using a multinomial sampling instead?

It avoids an auxilairy slice variable; by the virtue of Rao-Blackwell theorem, this can reduce variance in the samples and ensure better performance when computing expectations using HMC samples. 

> Does using a multinomial sampling from static trajectory make sense?

No, it's not a valid transition with the current static trajectory implementation https://github.com/TuringLang/AdvancedHMC.jl/pull/5/files#diff-f48defe9803ddea4f759bce857b70de0R8.